crates/asm/src/subprotocol/mod.rs

```diff
@@ -0,0 +1,47 @@
+//! Subprotocol trait definition for ASM.
+//!
+//! This trait defines the interface every ASM subprotocol implementation must
+//! provide. Each subprotocol is responsible for parsing its transactions,
+//! updating its internal state, and emitting cross-protocol messages and logs.
+
+use bitcoin::Block;
+use strata_primitives::buf::Buf32;
+
+use crate::{msg::InterProtoMsg, state::SectionState};
+
+/// Interface for ASM subprotocol implementations.
+///
+/// Each subprotocol must specify a unique `VERSION` and define the
+/// associated `State`, `Msg`, `TxTag`, and `AuxInput` types. Processing
+/// occurs in two phases: `process_block_txs` (initial pass) and
+/// `finalize_state` (after inter-protocol messaging).
+pub trait Subprotocol {
+    /// 1-byte subprotocol identifier / version tag (matches SPS-50).
+    const VERSION: u8;
+
+    /// Returns the identifier of the subprotocol for this section.
+    ///
+    /// This ID corresponds to the version or namespace of the subprotocol whose
+    /// state is serialized in this section.
+    fn id(&self) -> u8 {
+        Self::VERSION
+    }
+
+    /// Process the L1Block and extracts all the relevant information from L1 for the subprotocol
+    ///
+    /// Update it's own state and output a list of InterProtoMsg addressed to other subprotocols
+    fn process_block(&mut self, _block: &Block) -> Vec<(u8, InterProtoMsg)> {
+        vec![]
+    }
```

Comment:
We don't need a `process_block` function that takes the whole block, that introduces more complexity since we'd be parsing data out of the block in multiple places.  The subprotocol only sees txs that have its subprotocol ID.

Also, default impls are intended more for cases where it's a trivial extension of the existing logic like:

```rs
fn try_get_something(&self) -> Result<Option<T>, Error>;

fn expect_get_something(&self) -> T {
    self.try_get_something().unwrap().unwrap()
}
```

Or for how `clone_from` works.